### PR TITLE
Store parts tree result in Blob Storage

### PR DESCRIPTION
**Pull request by marcgs**
* Parts tree is stored in a blob container by the provider FlowController (provisioning done in [previous PR](https://github.com/catenax/tractusx/pull/272))
* Provider connector using a Filesystem Vault (FsVault)
* On job completion the blob is copied to the destination by the PartsTreeRecursiveJobHandler
* Status endpoint returns job status and, if status is COMPLETED, the SAS URL to access the blob
* Adapted integration and system tests

### CODE REVIEW

File: coreservices/partsrelationshipservice/connector/prs-connector-provider/src/main/java/net/catenax/prs/connector/provider/PartsRelationshipServiceApiToFileFlowController.java [hunk 1]
```diff
@@ -108,22 +123,38 @@ public DataFlowInitiateResponse initiateFlow(final DataRequest dataRequest) {
...
-            writeToFile(partRelationshipsWithInfos, Path.of(destinationPath));
-        } catch (IOException e) {
-            final String message = "Error writing file " + destinationPath + e.getMessage();
-            monitor.severe(message);
-            return new DataFlowInitiateResponse(ResponseStatus.FATAL_ERROR, message);
+        // Retrieve blob storage SAS token from vault
+        final var destSecretName = dataRequest.getDataDestination().getKeyName();
+        if (destSecretName == null) {
+            monitor.severe(format("No credentials found for %s, will not copy!", dataRequest.getDestinationType()));
+            return new DataFlowInitiateResponse(ResponseStatus.ERROR_RETRY, "Did not find credentials for data destination.");
```
Comment (algattik) [inline coreservices/partsrelationshipservice/connector/prs-connector-provider/src/main/java/net/catenax/prs/connector/provider/PartsRelationshipServiceApiToFileFlowController.java:130]: seems like a fatal error, not a RETRY one

File: coreservices/partsrelationshipservice/connector/prs-connector-provider/src/main/java/net/catenax/prs/connector/provider/PartsRelationshipServiceApiToFileFlowController.java [hunk 2]
```diff
@@ -108,22 +123,38 @@ public DataFlowInitiateResponse initiateFlow(final DataRequest dataRequest) {
...
-        // write to temporary file first, so that test does not pick up an empty file while writing
-        final var tmpPath = Path.of(path.getParent().toString(), format(".%s.tmp", path.getFileName()));
-        Files.writeString(tmpPath, content);
-        Files.move(tmpPath, path, REPLACE_EXISTING, ATOMIC_MOVE);
+    private void write(final DataAddress destination, final String blobName, final byte[] data, final String secretToken) {
+        final var containerName = destination.getProperty(AzureBlobStoreSchema.CONTAINER_NAME);
+        final var accountName = destination.getProperty(AzureBlobStoreSchema.ACCOUNT_NAME);
+        final var sasToken = typeManager.readValue(secretToken, AzureSasToken.class);
+
+        final var blobClient = new BlobClientBuilder()
```
Comment (algattik) [inline coreservices/partsrelationshipservice/connector/prs-connector-provider/src/main/java/net/catenax/prs/connector/provider/PartsRelationshipServiceApiToFileFlowController.java:145]: use BlobStoreApi.putBlob?

File: coreservices/partsrelationshipservice/docker-compose.yml [hunk 3]
```diff
@@ -97,6 +97,7 @@ services:
     volumes:
       - src-shared-volume:/tmp/copy/source
       - dst-shared-volume:/tmp/copy/dest
+      - ${edc_vault_certificate}:/cert.pfx
```
Comment (algattik) [inline coreservices/partsrelationshipservice/docker-compose.yml:100]: can delete this

File: .idea/misc.xml [hunk 4]
```diff
@@ -10,5 +10,8 @@
       </list>
     </option>
   </component>
+  <component name="PDMPlugin">
```
Comment (marcgs) [inline .idea/misc.xml:13]: revert this change

File: coreservices/partsrelationshipservice/connector/prs-connector-provider/dataspaceconnector-configuration.properties [hunk 5]
```diff
@@ -1 +1,6 @@
 web.http.port=8181
+edc.vault.clientid=0dcdabb0-70db-4ada-8a6b-20cbb6519824
+edc.vault.tenantid=836f6094-147a-44cc-8064-3dd900228759
+edc.vault.certificate=/tmp/cert.pfx
```
Comment (marcgs) [inline coreservices/partsrelationshipservice/connector/prs-connector-provider/dataspaceconnector-configuration.properties:4]: Not required if `FsVault` in place

File: coreservices/partsrelationshipservice/connector/prs-connector-consumer/pom.xml [hunk 6]
```diff
@@ -117,5 +127,11 @@
       <version>${revision}</version>
       <scope>compile</scope>
     </dependency>
+      <dependency>
+          <groupId>net.catenax.prs</groupId>
+          <artifactId>prs-client</artifactId>
+          <version>0.0.1-SNAPSHOT</version>
```
Comment (marcgs) [inline coreservices/partsrelationshipservice/connector/prs-connector-consumer/pom.xml:133]: extract variable?

File: coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/service/ConsumerService.java [hunk 7]
```diff
@@ -80,7 +112,27 @@ public JobInitiateResponse initiateTransfer(final FileRequest request) {
...
+                response.sasToken(createSasUrl(job.getJobData()).toString());
+            }
             return response.build();
         });
     }
+
+    private URL createSasUrl(final Map<String, String> jobData) {
+        final var containerName = jobData.get(CONTAINER_NAME_KEY);
+        final var destinationPath = jobData.get(DESTINATION_PATH_KEY);
+        if (containerName == null || destinationPath == null) {
```
Comment (marcgs) [inline coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/service/ConsumerService.java:125]: @algattik Should we use `Objects.requireNonNull` here? The EDC Framework uses this method to validate params, and we do as well in some cases.

Comment (algattik) [reply coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/service/ConsumerService.java:125]: I assumed we wanted to throw EdcException, but if the framework does as well, we sure can

File: coreservices/partsrelationshipservice/connector/prs-connector-provider/src/main/java/net/catenax/prs/connector/provider/BlobStorageClient.java [hunk 8]
```diff
@@ -0,0 +1,126 @@
...
+     * Factory for BlobClient
+     */
+    private final BlobClientFactory blobClientFactory;
+
+    /**
+     * @param monitor  Logger
+     * @param jsonUtil Type manager
+     * @param vault    Vault
+     */
+    public BlobStorageClient(final Monitor monitor, final JsonUtil jsonUtil, final Vault vault) {
```
Comment (algattik) [inline coreservices/partsrelationshipservice/connector/prs-connector-provider/src/main/java/net/catenax/prs/connector/provider/BlobStorageClient.java:53]: inline this

File: coreservices/partsrelationshipservice/connector/prs-connector-provider/src/test/java/net/catenax/prs/connector/provider/BlobStorageClientTest.java [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+        assertThat(new String(byteArrayInputStreamCaptor.getValue().readAllBytes())).isEqualTo(data);
+    }
+
+    @Test
+    public void writeToBlob_noSASToken_fails() {
+        // Arrange
+        when(vault.resolveSecret(keyName)).thenReturn(null);
+
+        // Act
+        assertThatThrownBy(() -> sut.writeToBlob(dataAddress, blobName, data)).hasMessage("Can not retrieve SAS token");
```
Comment (algattik) [inline coreservices/partsrelationshipservice/connector/prs-connector-provider/src/test/java/net/catenax/prs/connector/provider/BlobStorageClientTest.java:91]: specify exception type here.

nit: consider using assertThatExceptionOfType to align with other tests

File: coreservices/partsrelationshipservice/integration-tests/src/test/java/net/catenax/prs/systemtest/ConnectorSystemTests.java [hunk 10]
```diff
@@ -87,46 +81,61 @@ public void downloadFile() throws Exception {
...
-                });
+    private String getUrl(String sasUrl) throws IOException, InterruptedException {
+        var httpClient = HttpClient.newBuilder()
+                .version(HttpClient.Version.HTTP_1_1)
+                .build();
+
+        var request = HttpRequest.newBuilder()
+                .GET()
+                .uri(URI.create(sasUrl))
+                .setHeader("User-Agent", "Java 11 HttpClient Bot") // add request header
```
Comment (algattik) [inline coreservices/partsrelationshipservice/integration-tests/src/test/java/net/catenax/prs/systemtest/ConnectorSystemTests.java:120]:
```suggestion
```

File: coreservices/partsrelationshipservice/integration-tests/src/test/java/net/catenax/prs/systemtest/ConnectorSystemTests.java [hunk 11]
```diff
@@ -87,46 +81,61 @@ public void downloadFile() throws Exception {
                 .when()
                         .post("/api/v0.1/file")
                 .then()
+                        .log().all()
```
Comment (marcgs) [inline coreservices/partsrelationshipservice/integration-tests/src/test/java/net/catenax/prs/systemtest/ConnectorSystemTests.java:84]: remove rest assured log

File: coreservices/partsrelationshipservice/Dockerfile [hunk 12]
```diff
@@ -74,6 +74,13 @@ COPY --from=maven /build/connector/prs-connector-provider/target/prs-connector-p
 COPY --from=maven /build/connector/prs-connector-provider/src/main/resources/logging.properties .
 COPY --from=wget jmx-prometheus-agent.jar .
 COPY cd/jmx_prometheus_config.yml .
+
+# Settings for EDC FsVault (filesystem-based key vault)
+RUN cp /dev/null dataspaceconnector-vault.properties
```
Comment (ouphi) [inline coreservices/partsrelationshipservice/Dockerfile:79]: Add a comment to explain the command

File: coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/service/PartsTreeRecursiveJobHandler.java [hunk 13]
```diff
@@ -79,45 +95,65 @@
      */
     @Override
     public void complete(final MultiTransferJob job) {
-        monitor.info("Completed with recursive retrieval for Job " + job.getJobId());
+        monitor.info("Completed retrieval for Job " + job.getJobId());
+        final var completedTransfers = job.getCompletedTransfers();
+        final var accountName2 = configuration.getStorageAccountName();
```
Comment (Izzzu) [inline coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/service/PartsTreeRecursiveJobHandler.java:100]: Why `accountName2` and not just `accountName`? (Same for containerName2 and blobName2)

Comment (Izzzu) [reply coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/service/PartsTreeRecursiveJobHandler.java:100]: Wouldn't a name e.g `targetAccountName` be more descriptive? if it's related to the fact that it's the account to which the blob is going to be copied?

Comment (marcgs) [reply coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/service/PartsTreeRecursiveJobHandler.java:100]: I agree. I have renamed these variables.

File: coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/service/PartsTreeRecursiveJobHandler.java [hunk 14]
```diff
@@ -79,45 +95,65 @@
      */
     @Override
     public void complete(final MultiTransferJob job) {
```
Comment (Izzzu) [inline coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/service/PartsTreeRecursiveJobHandler.java:97]: nit: 
This method seemed to me at first a bit hard to read so I come up with some small refactoring to understand it better:
```java
@Override
    public void complete(final MultiTransferJob job) {
        monitor.info("Completed retrieval for Job " + job.getJobId());
        final var completedTransfers = job.getCompletedTransfers();
        final var accountName2 = configuration.getStorageAccountName();
        final var containerName2 = job.getJobData().get(ConsumerService.CONTAINER_NAME_KEY);
        final var blobName2 = job.getJobData().get(ConsumerService.DESTINATION_PATH_KEY);
        final byte[] blob = getBlob(completedTransfers);
        monitor.info(format("Putting blob to %s/%s/%s",
                accountName2,
                containerName2,
                blobName2
        ));
        blobStoreApi.putBlob(accountName2, containerName2, blobName2, blob);
    }

    private byte[] getBlob(List<TransferProcess> completedTransfers) {
        if (completedTransfers.isEmpty()) {
            final var result = new PartRelationshipsWithInfos();
            return jsonUtil.asString(result).getBytes(StandardCharsets.UTF_8);
        } else {
            final var firstTransfer = completedTransfers.get(0);
            final var destination = firstTransfer.getDataRequest().getDataDestination();
            final String accountName1 = destination.getProperty(AzureBlobStoreSchema.ACCOUNT_NAME);
            final String containerName1 = destination.getProperty(AzureBlobStoreSchema.CONTAINER_NAME);
            final String blobName1 = firstTransfer.getDataRequest().getProperties().get(PrsConnectorConstants.DATA_REQUEST_PRS_DESTINATION_PATH);
            monitor.info(format("Getting blob from %s/%s/%s",
                    accountName1,
                    containerName1,
                    blobName1
            ));
            return blobStoreApi.getBlob(accountName1, containerName1, blobName1);
        }
    }
```
With that we are separating the levels of abstractions per method: first we are getting the blob and then we are saving it. How the blob is fetched it's another level of abstraction so for me personally it's easier to read if we want to see only the upper-level flow (not going into many details if the completed transfers are empty or not).

But I'll leave it up to you to decide which version you want to keep.

Comment (algattik) [reply coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/service/PartsTreeRecursiveJobHandler.java:97]: Yeah that was a minimal implementation to get ITs working, the real one will come when we implement full distributed PRS

**Review by algattik — APPROVED** [review]

File: coreservices/partsrelationshipservice/Dockerfile [hunk 15]
```diff
@@ -74,6 +74,14 @@ COPY --from=maven /build/connector/prs-connector-provider/target/prs-connector-p
 COPY --from=maven /build/connector/prs-connector-provider/src/main/resources/logging.properties .
 COPY --from=wget jmx-prometheus-agent.jar .
 COPY cd/jmx_prometheus_config.yml .
+
+# Settings for EDC FsVault (filesystem-based key vault)
+# Creates an empty property file or empty an already existing file.
```
Comment (algattik) [inline coreservices/partsrelationshipservice/Dockerfile:79]:
```suggestion
# Creates an empty property file.
```

**Review by ouphi — APPROVED** [review]

File: coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/service/PartsTreeRecursiveJobHandler.java (hunk 14 above)
Comment (marcgs) [reply coreservices/partsrelationshipservice/connector/prs-connector-consumer/src/main/java/net/catenax/prs/connector/consumer/service/PartsTreeRecursiveJobHandler.java:97]: @Izzzu Given that we will rework it in the coming [PR](https://github.com/catenax/tractusx/pull/287) I would suggest not to put any more effort
